fix: do not set the commitIndex to be greater than biggest logIndex

entangled90 opened the pull request:
## Description
It may happen that a follower that is failing to flush entries will receive a heartbeat from the leaders (an AppendEntries with entries=[]). Because we don't flush on followers when changing the commitIndex, the follower will increase its commitIndex.
The next time the leader tries to send an AppendEntriers to the follower, if it's still unable to flush, it will try to reset the log. However, because now commitIndex> last logIndex, it will fail to reset and will become inactive.

The fix is to never set the local `commitIndex` to be greater than the last persisted log entry index.

## Related issues

closes #27852

entangled90 committed:
fix: do not set the commitIndex to be greater than biggest logIndex

It may happen that a follower that is failing to flush entries will
receive a heartbeat from the leaders (an AppendEntries with entries=[]).
Because we don't flush on followers when changing the commitIndex, the
follower will increase its commitIndex.
The next time the leader tries to send an AppendEntriers to the
follower, if it's still unable to flush, it will try to reset the log.
However, because now commitIndex> last logIndex, it will fail to reset
and will become inactive.

entangled90 committed:
refactor: merge checks on commitIndex/raftLog.getLastIndex()

entangled90 commented on zeebe/atomix/cluster/src/main/java/io/atomix/raft/impl/RaftContext.java:
```diff
@@ -546,8 +546,14 @@ public void notifyApplicationEntryCommittedPositionListeners(final long committe
+    // Do not set the commitIndex to be greater than the events we persisted.
+    // In case of a heartbeat, the leader sends its commitIndex which may be greater than what's
+    // persisted in case we failed to persist a previous AppendEntries
+    commitIndex = Math.min(commitIndex, raftLog.getLastIndex());
```
What I noticed in the logs is that we received an heartbeat with` leader.commitIndex >  follower.commitIndex`.
In `checkPreviousEntry` we check:
- prevLogIndex <= log.lastEntryIndex
- if prevLogIndex < log.LastEntryIndex:
  - all entries in the log must match (same index, same term)
- request.term == lastEntryTerm

For reference, this is the appendEntries request that caused the issue:
`InternalAppendRequest[term=1, leader=2, prevLogIndex=0, prevLogTerm=0, commitIndex=3, entries=[ 3 entries]}`
For this particular case :
```java
  protected void appendEntries(
      final InternalAppendRequest request, final CompletableFuture<AppendResponse> future) {
    // Compute the last entry index from the previous log index and request entry count.
    final long lastEntryIndex = request.prevLogIndex() + request.entries().size();

    // Ensure the commitIndex is not increased beyond the index of the last entry in the request.
    final long commitIndex = Math.min(request.commitIndex(), lastEntryIndex);
```
commitIndex will be `3`, even though `prevLogIndex = 0`

lenaschoenburg replied:
Okay, based on the code comment I assumed it was a heartbeat (without any entries), not a real append. Shouldn't this follower then try to append the entries to the log and flush? If the flush succeeds, we can update the commit index to match. If not, we should not update the commit index. So perhaps the fix should rather be that we only update the commit index after flush, not before?

entangled90 replied:
For regular appendEntries, commitIndex is updated only after successful flushes.

However, for heartbeats, there is no entry to flush to the journal, so the commitIndex is updated without flushing first the journal (we write only in the `MetaStore`)

entangled90 replied:
oh you are right, got confused :confused: 
Then we should definitely move the `setCommitIndex` after the flush.
I would still keep the check in the setter as a last resort to avoid strange bugs.

entangled90 commented:
I'll try to spend a bit more time on trying to replicate the issue deterministically :+1:

entangled90 committed:
fix: setCommitIndex only after flushing all the entries

Set the commitIndex of the node only after successfully flushing the
entries.
The commitIndex stored in the node should not be greater than the
lastLogIndex. Without flushing first, the lastLogIndex might not
actually be persisted successfully, so the node may refuse to reset the
log to a previous position even though it hasn't actually persisted that
data.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/camunda/camunda?pullRequest=30354) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/camunda/camunda?pullRequest=30354) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/camunda/camunda?pullRequest=30354) it.</sub>

entangled90 commented:
Funnily enough test #27852 has not been reported as flaky for a couple of days now

ChrisKujawa commented:
I guess this is something we still want to do? @entangled90 @lenaschoenburg ?

deepthidevaki reviewed (approved):
LGTM. The fix is that the commit index is only updated after flush. right?. This makes sense. It was not an issue before, because commitIndex was just stored in memory. 

For testing, we can probably verify directly that the faulty nodes do not update commitIndex until it is not faulty anymore, using faultynode's `RaftContext.getCommitIndex()). Is that enough?

entangled90 commented:
I think the issue started arising when we added checks on deleting entries in the log based on the commitIndex. 
From then onwards, the commitIndex must reflect what's inside the log as well, otherwise strange exceptions may happen

backport-action commented:
Created backport PR for `stable/8.5`:
- #38116 with remaining conflicts!

Please cherry-pick the changes locally and resolve any conflicts.
```bash
git fetch origin backport-30354-to-stable/8.5
git worktree add --checkout .worktree/backport-30354-to-stable/8.5 backport-30354-to-stable/8.5
cd .worktree/backport-30354-to-stable/8.5
git reset --hard HEAD^
git cherry-pick -x b494f4b41eb3d4bac689059e88828c4a97cb708c a70d46c31516b3c452bbc35f60d0441ad591416c
git push --force-with-lease
```

backport-action commented:
Successfully created backport PR for `stable/8.8`:
- #38119
